### PR TITLE
[Array API] Add linalg.vecdot

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #70542

This PR adds the function `linalg.vecdot` specified by the [Array
API](https://data-apis.org/array-api/latest/API_specification/linear_algebra_functions.html#function-vecdot)

For the complex case, it chooses to implement \sum x_i y_i. See the
discussion in https://github.com/data-apis/array-api/issues/356

Edit. When it comes to testing, this function is not quite a binopt, nor a reduction opt. As such, we're this close to be able to get the extra testing, but we don't quite make it. Now, it's such a simple op that I think we'll make it without this.

Resolves https://github.com/pytorch/pytorch/issues/18027.

cc @mruberry @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi

### CODE REVIEW

File: torch/linalg/__init__.py [hunk 1]
```diff
@@ -2251,3 +2251,42 @@
     >>> torch.dist(Q.mT @ Q, torch.eye(4))
     tensor(6.2158e-07)
 """)
+
+vecdot = _add_docstr(_linalg.linalg_vecdot, r"""
+linalg.vecdot(x, y, *, dim=-1, out=None) -> Tensor
+
+Computes the vector product of two batches of vectors along a dimension.
```
Comment (IvanYashchuk) [inline torch/linalg/__init__.py:2258]: There are several "vector products", this one computes the vector _dot_ product

File: torch/linalg/__init__.py [hunk 2]
```diff
@@ -2251,3 +2251,42 @@
...
+
+In symbols, this function computes
+
+.. math::
+
+    \sum_{i=1}^n x_iy_i.
+
+over the dimension :attr:`dim`.
+
+Supports input of half, bfloat16, float, double, cfloat, cdouble and integral dtypes.
```
Comment (IvanYashchuk) [inline torch/linalg/__init__.py:2268]: If it supports everything do we need to mention it?

I think it's fine to be explicit here. Also, many other linalg functions have a note on supported dtypes.

File: torch/linalg/__init__.py [hunk 3]
```diff
@@ -2251,3 +2251,42 @@
...
+
+        :func:`torch.matmul` computes a general matrix-matrix multiplication for batches
+        of matrices.
+
+Args:
+    x (Tensor): first batch of vectors.
+    y (Tensor): second batch of vectors.
+
+Keyword args:
+    dim (int): Dimension along which to compute the vector product.
```
Comment (IvanYashchuk) [inline torch/linalg/__init__.py:2281]:
```suggestion
    dim (int): Dimension along which to compute the vector product. Default: `-1`.
```

**Review by IvanYashchuk — APPROVED** [review]
Great!

This PR also resolves https://github.com/pytorch/pytorch/issues/18027.

File: torch/linalg/__init__.py (hunk 2 above)
Comment (Lezcano) [reply torch/linalg/__init__.py:2268]: Leaving it as "better explicit than implicit". Also, it does support `bool`, but it's not announced here. If I had to choose I'd disallow integral and boolean types, but left them because why not.

File: torch/linalg/__init__.py (hunk 1 above)
Comment (Lezcano) [reply torch/linalg/__init__.py]: Yup. The only thing is that this PR implements \sum x_i y_i, which is not the complex dot product. I opened an issue here https://github.com/data-apis/array-api/issues/356. For now, I'm calling this "the real dot product", which is not standard notation either.

I don't think that there's a standard name for this in the complex case. You'd call it C-linear complex dot product (perhaps) to differentiate it from the R-linear real dot product `Re(x^H y)`...

Leaving it like this for now.

File: torch/_torch_docs.py [hunk 4]
```diff
@@ -3429,21 +3429,34 @@ def merge_dicts(*dicts):
...
+vectors, and it is the identity for real vectors.
 
 .. note::
 
     Unlike NumPy's vdot, torch.vdot intentionally only supports computing the dot product
     of two 1D tensors with the same number of elements.
 
+.. seealso::
+
+        :func:`torch.linalg.vecdot` computes the dot product of two batches of vectors along a dimension.
```
Comment (mruberry) [inline torch/_torch_docs.py:3450]: Nice seealso addition

File: torch/linalg/__init__.py [hunk 5]
```diff
@@ -2251,3 +2251,38 @@
...
+    \sum_{i=1}^n \overline{x_i}y_i.
+
+over the dimension :attr:`dim` where :math:`\overline{x_i}` denotes the conjugate for complex
+vectors, and it is the identity for real vectors.
+
+Supports input of half, bfloat16, float, double, cfloat, cdouble and integral dtypes.
+It also supports broadcasting.
+
+Args:
+    x (Tensor): first batch of vectors.
```
Comment (mruberry) [inline torch/linalg/__init__.py:2273]: If this supports batching should the main doc be updated with a sentence on batching, too?

Comment (mruberry) [reply torch/linalg/__init__.py:2273]: (It's mentioned briefly in the initial sentence describing the operation, but like, what output shape should the user expect? Is it obvious?)

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -1101,7 +1101,7 @@ class ReductionOpInfo(OpInfo):
     the optional keyword parameters of the ReductionOpInfo constructor.
 
     If a reduction operator does not yet implement the full required API of
-    reduction operators, this should be documented by skipping the failing
+    reduction operators, this should be documented by xfailing the failing
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:1104]: Nice update

File: torch/testing/_internal/common_methods_invocations.py [hunk 7]
```diff
@@ -2012,6 +2012,15 @@ def sample_inputs_isclose(
         yield SampleInput(lhs, args=(rhs,),
                           kwargs=dict(op_kwargs, rtol=rtol, atol=atol, equal_nan=equal_nan))
 
+def sample_inputs_linalg_vecdot(op_info, device, dtype, requires_grad, **kwargs):
+    yield from sample_inputs_binary_pwise(op_info, device, dtype, requires_grad)
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:2016]: Not in love with reusing the elementwise binary sample inputs for this op -- wouldn't it be simple to generate some examples with and without batch dims, an empty tensor, and dim of the default, -2, 1, and 0?

File: torch/testing/_internal/common_methods_invocations.py [hunk 8]
```diff
@@ -9778,6 +9787,16 @@ def ref_pairwise_distance(input1, input2):
            gradcheck_wrapper=gradcheck_wrapper_hermitian_input,
            decorators=[skipCUDAIfNoMagmaAndNoCusolver, skipCUDAIfRocm, skipCPUIfNoLapack],
            ),
+    OpInfo('linalg.vecdot',
+           aten_name='linalg_vecdot',
+           ref=lambda x, y, *, dim=-1: (x.conj() * y).sum(dim),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:9792]: Nice reference

**Review by mruberry — COMMENTED** [review]
Overall looks great, @Lezcano. I left a few inline comments -- the one change I think we want to make is to just write a few sample inputs for this op specifically, especially because it doesn't have any tests beyond OpInfo tests. Just ping me when this ready to merge

**Comment by Lezcano** [comment]
This one should be ready @mruberry.

**Comment by Lezcano** [comment]
Yes, we should land this one. It was a bit left out due to the whole primtorch effort, but we should still land it.

**Comment by Lezcano** [comment]
I just rebased the PR. If the tests go green, I'd say we merge it.

**Review by mruberry — APPROVED** [review]
Ship it!

**Comment by malfet** [comment]
@Lezcano linalg PRs often cause CUDA-10.2 regressions, can you please test for it before submitting changes?

**Comment by Lezcano** [comment]
The CI was green, so I assumed everything was fine. Should I run PRs on `trunk` or how should I test them? @malfet
